core/src/View/Kokkos_BasicView.hpp

```diff
@@ -0,0 +1,611 @@
+  KOKKOS_DEFAULTED_FUNCTION constexpr BasicView(BasicView &&)      = default;
+
+  KOKKOS_FUNCTION constexpr BasicView(const mdspan_type &other)
+      : ptr(other.data_handle()), map(other.mapping()), acc(other.accessor()){};
+  KOKKOS_FUNCTION constexpr BasicView(mdspan_type &&other)
+      : ptr(std::move(other.data_handle())),
+        map(std::move(other.mapping())),
+        acc(std::move(other.accessor())){};
+
+  template <class... OtherIndexTypes>
+  //    requires(std::is_constructible_v<mdspan_type, data_handle_type,
+  //                                     OtherIndexTypes...>)
+  KOKKOS_FUNCTION explicit constexpr BasicView(
+      std::enable_if_t<std::is_constructible_v<mdspan_type, data_handle_type,
+                                               OtherIndexTypes...>,
+                       data_handle_type>
+          p,
+      OtherIndexTypes... exts)
+      : ptr(std::move(p)),
+        map(extents_type(static_cast<index_type>(std::move(exts))...)),
+        acc{} {}
+
+  template <class OtherIndexType, size_t Size>
+  // When doing C++20 we should switch to this, the conditional explicit we
+  // can't do in 17
+  //    requires(std::is_constructible_v<mdspan_type, data_handle_type,
+  //                                     std::array<OtherIndexType, Size>>)
+  //  explicit(Size != rank_dynamic())
+  KOKKOS_FUNCTION constexpr BasicView(
+      std::enable_if_t<
+          std::is_constructible_v<mdspan_type, data_handle_type,
+                                  std::array<OtherIndexType, Size>>,
+          data_handle_type>
+          p,
+      const Array<OtherIndexType, Size> &exts)
+      : ptr(std::move(p)), map(extents_type(exts)), acc{} {}
+
+  KOKKOS_FUNCTION constexpr BasicView(data_handle_type p,
+                                      const extents_type &exts)
+// Compilation will simply fail in C++17 and overload set should not be an issue
+#ifdef KOKKOS_ENABLE_CXX20
+    requires(std::is_default_constructible_v<accessor_type> &&
+             std::is_constructible_v<mapping_type, const extents_type &>)
+#endif
+      : ptr(std::move(p)), map(exts), acc{} {
+  }
+
+  KOKKOS_FUNCTION constexpr BasicView(data_handle_type p, const mapping_type &m)
+// Compilation will simply fail in C++17 and overload set should not be an issue
+#ifdef KOKKOS_ENABLE_CXX20
+    requires(std::is_default_constructible_v<accessor_type>)
+#endif
+      : ptr(std::move(p)), map(m), acc{} {
+  }
+
+  KOKKOS_FUNCTION constexpr BasicView(data_handle_type p, const mapping_type &m,
+                                      const accessor_type &a)
+      : ptr(std::move(p)), map(m), acc(a) {}
+
+  template <class OtherT, class OtherE, class OtherL, class OtherA>
+//    requires(std::is_constructible_v<mdspan_type,
+//                                     typename BasicView<OtherT, OtherE,
+//                                     OtherL,
+//                                                        OtherA>::mdspan_type>)
```

Comment:
Why are some of the requires-clauses commented and others guarded with `KOKKOS_ENABLE_CXX20`?

Reply:
Because the ones I guarded can't be implemented with SFINAE. But I want the constraint behavior at least in C++20. For the ones where I commented them out the SFINAE works in C++20 too, so I thought its better to not have ifdefs multiple disjointed lines, and just left the comment in there so we a) can read easily what was the intend, and b) already have the lines we want when we get rid of C++17.